// @umijs/preset-umi/dist/features/depsOnDemand/depsOnDemand.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/features/depsOnDemand/depsOnDemand.ts
var depsOnDemand_exports = {};
__export(depsOnDemand_exports, {
  default: () => depsOnDemand_default
});
module.exports = __toCommonJS(depsOnDemand_exports);
var import_utils = require("@umijs/utils");
var import_fs = require("fs");
var import_path = require("path");
var depsOnDemand_default = (api) => {
  api.onStart(() => {
    var _a, _b;
    const hasSwcConfig = api.config.srcTranspiler === "swc" || api.config.depTranspiler === "swc";
    const swcInstalled = ((_a = api.pkg.dependencies) == null ? void 0 : _a["@swc/core"]) || ((_b = api.pkg.devDependencies) == null ? void 0 : _b["@swc/core"]);
    if (hasSwcConfig && !swcInstalled) {
      api.logger.info("Since swc is used, install @swc/core on demand.");
      addDeps({
        pkgPath: api.pkgPath || (0, import_path.join)(api.cwd, "package.json"),
        name: "@swc/core",
        version: `^${require("@umijs/bundler-webpack/package.json").devDependencies["@swc/core"]}`
      });
      (0, import_utils.installWithNpmClient)({
        npmClient: api.appData.npmClient,
        cwd: api.cwd
      });
    }
  });
};
function addDeps(opts) {
  const pkg = (0, import_fs.existsSync)(opts.pkgPath) ? import_utils.fsExtra.readJSONSync(opts.pkgPath) : {};
  pkg.devDependencies = pkg.devDependencies || {};
  pkg.devDependencies[opts.name] = opts.version;
  import_utils.fsExtra.writeJSONSync(opts.pkgPath, pkg, { spaces: 2 });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
