// @umijs/preset-umi/dist/commands/generators/component.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/commands/generators/component.ts
var component_exports = {};
__export(component_exports, {
  ComponentGenerator: () => ComponentGenerator,
  default: () => component_default
});
module.exports = __toCommonJS(component_exports);
var import_core = require("@umijs/core");
var import_utils = require("@umijs/utils");
var import_path = require("path");
var import_constants = require("../../constants");
var import_utils2 = require("./utils");
var component_default = (api) => {
  api.describe({
    key: "generator:component"
  });
  api.registerGenerator({
    key: "component",
    name: "Generate Component",
    description: "Generate component boilerplate code",
    type: import_core.GeneratorType.generate,
    fn: async (options) => {
      const h = new import_utils2.GeneratorHelper(api);
      options.generateFile;
      let componentNames = options.args._.slice(1);
      if (componentNames.length === 0) {
        let name = "";
        name = await h.ensureVariableWithQuestion(name, {
          type: "text",
          message: "Please input you component Name",
          hint: "foo",
          initial: "foo",
          format: (s) => (s == null ? void 0 : s.trim()) || ""
        });
        componentNames = [name];
      }
      for (const cn of componentNames) {
        await new ComponentGenerator({
          srcPath: api.paths.absSrcPath,
          appRoot: api.paths.cwd,
          generateFile: import_utils.generateFile,
          componentName: cn
        }).run();
      }
    }
  });
};
var ComponentGenerator = class {
  constructor(opts) {
    this.opts = opts;
    const { name, dir } = (0, import_path.parse)(this.opts.componentName);
    this.name = name;
    this.dir = dir;
  }
  async run() {
    const { generateFile: generateFile2, appRoot } = this.opts;
    const capitalizeName = import_utils.lodash.capitalize(this.name);
    const base = (0, import_path.join)(this.opts.srcPath, "components", this.dir, capitalizeName);
    const indexFile = (0, import_path.join)(base, "index.ts");
    const compFile = (0, import_path.join)(base, `${capitalizeName}.tsx`);
    await generateFile2({
      target: indexFile,
      path: INDEX_TPL,
      baseDir: appRoot,
      data: { compName: capitalizeName }
    });
    await generateFile2({
      target: compFile,
      path: COMP_TPL,
      baseDir: appRoot,
      data: { compName: capitalizeName }
    });
  }
};
var INDEX_TPL = (0, import_path.join)(import_constants.TEMPLATES_DIR, "generate/component/index.ts.tpl");
var COMP_TPL = (0, import_path.join)(import_constants.TEMPLATES_DIR, "generate/component/component.tsx.tpl");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ComponentGenerator
});
